### ISSUE TITLE
[Bug]: logical operators do not work on nilable classes

### ISSUE THREAD

**Issue by jabraham17** [chapel-lang/chapel#25382]
### Summary of Problem

**Description:**
Using logical operators (`&&` and `||`) on a nilable class causes strange resolution error that it cannot match `isTrue`. Its not clear to me if this is intended to work since we explicitly disallow `&&` and `||` on integral types.

Making it work is an easy patch, add `inline proc isTrue(x:borrowed class?) do return x != nil;` to `ChapelBase`. Making it a better error message is also an easy patch, add `proc isTrue(x:borrowed class?) do compilerError("short-circuiting logical operators not supported on classes");`

Note there is a similar issue with `c_ptr`, which gives an error about not being able to promote `&&`/`||`. I believe the `c_ptr` case can be solved similarly.

**Is this issue currently blocking your progress?**
no, can be worked around with an explict `!= nil`

### Steps to Reproduce

**Source Code:**

```chapel
class C { }
var x = new unmanaged C?();
var y = new unmanaged C?();
// if x { } // this is fine
if x && y { } // error: unable to resolve call isTrue
if x || y { } // error: unable to resolve call isTrue
```

```chapel
use CTypes;
var x: c_ptr(void) = nil;
var y = x;
// if x { } // this is fine
if x && y { } // error: cannot promote short-circuiting &&
```

**Compile command:**
`chpl foo.chpl`

### Configuration Information

- Output of `chpl --version`: chpl version 2.1.0 pre-release
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: darwin
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: arm64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads *
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: system *
CHPL_HWLOC: system *
CHPL_RE2: bundled *
CHPL_LLVM: system *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: LLVM 15.0.7

**Comment by dlongnecke-cray** [comment]
Here's the OP I was writing when I ran into this before I wisely decided to check to see if an existing issue was opened (thanks Jade!):

---

### Summary of Problem

The following program emits compiler errors:

```chapel
proc main() {
  var e: owned Error? = new Error('Secret test!');

  // OK.
  if e then writeln('OK!');

  // Error: unresolved call 'isTrue(owned Error?)'
  if e && e!.message() == 'Secret test!' then writeln('Gotcha!');
}
```

Since the idiom `if e then ...` works, it seems reasonable to expect that `if e && e!.message()` would also work (this is often a pattern I write in C++ when working with pointers, for example).

However, I get a compiler error complaining about no valid overload of `isTrue`. I believe this is an internal procedure used to help implement short-circuiting operators e.g., `&&`.

Should `&&` work when one or both of the actuals are class types?

While here, a related question is whether or not `&&` should be implemented using `isTrue` at all, or if it should be using a different internal procedure such as `_cond_test`. That procedure is more robust and includes even more errors for when different actual types don't make sense.

**Comment by DanilaFe** [comment]
This is because, unlike all other conditional operators I know of (like `if`), which use `_cond_test`, the lowering for `&&` and `||` uses `isTrue`. `isTrue` is only used in desugaring short-circuited conditionals, and nowhere else. It is explicitly defined only on booleans, and not even on integers.:

https://github.com/chapel-lang/chapel/blob/b16cae726c937ef27595ef7f76198f15ec14ef3d/modules/internal/ChapelBase.chpl#L1047-L1050

This definition predates our Git history. I think it's still possible to trace it back to the initial introduction of the code (to see what motivated it), but I haven't done so.

On the other hand, `_cond_test` (which is what's used in @dlongnecke-cray's `if e then` case) has support for `c_ptr` (as mentioned in @jabraham17's original OP) and any classes, including `Error`. 

https://github.com/chapel-lang/chapel/blob/b16cae726c937ef27595ef7f76198f15ec14ef3d/modules/internal/ChapelBase.chpl#L1363-L1364

Since uses of `isTrue` are limited to `||` and `&&`, and since it makes sense to allow `if e && bla` if we allow `if e`, I think the reasonable approach would be to remove `isTrue` altogether and use `_cond_test`. This would also improve error messages in strange cases, such as `if iterableExpr` and even `[1,2,3]` (which would get turned into `[x in [1,2,3]] isTrue(x)` by short circuiting based on my cursory reading).

However, existing code did explicitly disallow integers (and, presumably also other types) from being used in short-circuiting, so making this change would likely require dev. team consensus.

**Comment by bradcray** [comment]
I think from the historical perspective, there was some hesitancy to support `if expr` for non-bool `expr` expressions, but that we eventually prevailed and got it in.  I think that the hesitancy about extending it to permit `if expr && expr` for non-bool `expr`s may have been to avoid confusion between `if myInt && yourInt` and `if myInt & yourInt`, but I'm not confident about that memory.  I generally agree that if something is legal in a conditional, it should probably be legal in an `&&`, `||`, or `!` expression.

I think `isTrue()` as a standalone routine was introduced in https://github.com/chapel-lang/chapel/commit/e180b335b408a169ecb82a149c5aea19e3b310d3 and that prior to that it had been a method, introduced in https://github.com/chapel-lang/chapel/commit/e7743aceebc68aa1eb11d6df4cfe01c5f3b596a9 (?)

**Comment by DanilaFe** [comment]
We held a good discussion about whether this feature should be supported in the language during a [project meeting](https://github.com/chapel-lang/chapel/discussions/28252). My general summary:

We would be changing `&&` such that its type is roughly `&& : boolish x boolish -> bool`, where `boolish` are values that can be used in conditionals today. This change leaves the operator closed, because `bool` is truthy/`boolish`. Chaining `a && b && c` continues to work "for free".
* _Note:_ This is a restricted version of `&&`  compared to some other languages, where `&&` desugars roughly to `if a then b else a`. In those languages, the signature has the type `&& : A x B -> A | B`, where `A` and `B` are `boolish`. We can't type this because we don't have type unions.

The current mechanism enables custom types to be made `boolish`, via defining `chpl_cond_test_method`. Today, this is not user-facing and used only by the `Regex` module. In theory, we could come up with a mechanism for a user to turn their types into `bool`-able things. 

Historically, we've been shy about using integers even in regular boolean contexts (like `if 1 {}`), but we have warmed up to it. However, at the time, part of the shyness for `&&` was to avoid confusion with `&`, where a user might've been trying to write a bitwise expression (`0b11 & 0b10`). However, @bradcray mentioned that he feels fine about this if it's how other languages behave. See the table below for a survey of languages where both operators are present, w.r.t. to he semantics of `&&`.

Another concern was whether this opens a can of work w.r.t. implicit coercions. I've determined that C, C++, and Objective-C all inherit their "boolish" rules from C, which originally didn't have booleans at all. These languages are outliers in their permissiveness; in addition to allowing integers and pointers in `if` and `&&`, they also allow implicit coercions of these to `bool` formals. The `operator bool` I've described in the linked discussion turns out to be a [user-defined conversion function](https://en.cppreference.com/w/cpp/language/cast_operator.html), which matches the built-in behavior and enables implicit coercion into `bool` parameters. **Every other language surveyed does not permit this.** In my opinion, if we enable `int` and `class` to be used in `&&` in Chapel, we **would not enable general implicit coercions for these types to `bool`**. This is matched by the proposed implementation approach. 

| Language with bitwise and logical 'and'/'or' | Allows boolish in `if` | Allows boolish in `&&` | Auto-coerces boolish to `bool` parameters|
|--------------------------------------------|-----------------------|-----------------------|----------------------------------------------|
| C                                                                     | ✅ (no bools) | ✅ (no bools) | ✅  (no bools) |
| C++* ([ex](https://ato.pxeger.com/run?1=fZA9CsJAEIWxEuYUjwghKWyChZjoKSzsZN2sYUmyK9ldCUhOYpPG0gt5GhOSqJXF_DHzPZh3f_JL4Uwfx4zztn04e16uX7P5QipeuFQgkdrYSrByR-SMVBkUK4W5MC5gbBoTXbVMYVkuzPGkdRH0CazKQtwI4NpZJAm8vdb5Bl7fd8u-CJUWMTVEUlmUTKpgQJizGjW2WEVxN8ozgnrY_MgdIA1s5YT3kQK6k2Yi4Pv4gyGIwi86gj9f1GEMagY_Rlsme94))                                         | ✅                   | ✅                  | ✅                    |
| Objective-C*                                                  | ✅                   | ✅                  | ✅                    |
| Rust                                                               | ❌                  | ❌                   | ❌                   |
| Swift                                                              | ❌                  | ❌                   | ❌                   |
| Crystal  ([ex](https://ato.pxeger.com/run?1=m708uaiyuCQxZ8GCpaUlaboWNwtSUtMUUvLji_NzU0syMvPSNSoUrBSc8vNzNLlS81K4uCoUbBVMjLgy0xQquBQUCkpLihWUQopKUxWVIPIgCQU1NTRZBQ0jTZgKZNMVKiD2Qq2HOQMA))                                                    | ✅                   | ✅                  | ❌                   |
| TypeScript  ([ex](https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABFAhgawKYGcBCc4A2AFAB4BciARvgRimAJSIDeAvgFDsQJZSImIAvIgAsAJgDc7GMESkmzdokTcwWQhgB0BOAHMiAIgAaiGFmQAnEBgCEBhlI4y5AgGSv+CpSp4btewxMzS2s5MQY7B3YOVExcGnkJIA))                                                    | ✅                   | ✅                  | ❌                   |
| Chapel  (`main`)                                                    | ✅                   | ❌                   | ❌                   |
| Chapel  (as proposed here, [possible implementation](https://github.com/chapel-lang/chapel/pull/28102))                                                    | ✅                   | ✅                    | ❌                   |

* = Inherits C rules for `&&` etc.

Above, it's clear that Chapel is an exception; languages that permit boolish things in `if` also permit them in `&&`. Implementing `&&` would resolve this inconsistency.

Some team members have expressed concern for code like:

```Chapel
var x = 1 && 2;
```

Where the short-circuiting operator is used in a general expression context. Once again, in the above table, languages that permit non-`bool`s in `if` also permit them to be used in expression-level contexts. Viewed this way, it _looks_ like we are permitting arbitrary coercions to `bool` (since `&&` might be written as `operator &&(lhs: bool, rhs: bool)`.

However, `&&` is not a "standard" operator. It already has semantics that differ from normal expressions. In particular, this is the behavior of short-circuiting, in which case the RHS is not evaluated if it is not needed. Thus, if we were to allow `&&` to take non-booleans, we would be counting "interprets X as boolean" as another special-behavior that does not extend to other operators.

Finally, there are patterns that benefit from this working. For classes, I think the cleanest example is a safe method call:

```Chapel
var x: SomeClass?
if x && x!.someCondition() {

}
```

A similar idea applies to `Regex` matches.

```Chapel
if match && str[match.byteOffset] { 

}
```

**Comment by bradcray** [comment]
@DanilaFe : I wonder if you'd consider adding a line to your table summarizing `chapel (as proposed here)` as "check, check, X"?

**Comment by e-kayrakli** [comment]
My 2 cents is that consistency important, and I am OK with the proposal.

If it was a snap of a finger, I would make Chapel "X, X, X" in the table above, both because that's way more intuitive to me (I think this conundrum in programming exists because C didn't have bools, and we got used to it), and if I was to align Chapel with a group of languages in the table, I would align it with Rust and Swift. That's why I am not necessarily saying "of course, we have to do this, how did we miss it!". At the same time, I am not motivated or bothered enough by the first check mark to argue for a bigger change myself.

**Comment by bradcray** [comment]
I like the proposal here and would consider it an improvement to where we are today.  I wouldn't be a champion for switching to X, X, X, in Chapel n.0, but also wouldn't necessarily be an opponent if someone wanted to be that champion, seeing alignment with Swift, and to an extent Rust, as a good thing.  I think if someone did want to champion X, X, X, that should get an issue of its own.

**Comment by jabraham17** [comment]
I like this proposal of "check, check, X" for Chapel.

While ideally I wouldn't want to allow integers in conditionally (so "X, X, X"), I do want it for classes (nil-ability checks) and regex matches. I find it hard to rationalize this inconsistency, so I like "check, check, X".

Something I want to double check, the statement "Auto-coerces boolish to bool parameters" applies to assignment/initialization correct? So if Chapel does not support "Auto-coerces boolish to bool parameters", then the following is an error

```chapel
var x: bool;
x = 1; // illegal, no coercion to bool
```

the only coercion occurs in conditions/loops, `&&`, and `||`.

**Comment by DanilaFe** [comment]
Yes, that's correct. `var x: bool; x = 1` is not allowed today and this proposal does not permit it.

**Comment by arifthpe** [comment]
I weakly prefer the status quo for Chapel.
- I don't view it as inconsistent to allow bool-ish in `if` but not in `&&`/`||`; `if` conditions allowing bool-ish is already a special case and there's nothing wrong with not extending that special case.
- I don't want to allow things like `var x = 1 && 2;`, or worse `var x = 1; x &&= 2;`. I can't think of a practical issue with allowing this, it just feels bad.
- I see the point that `&&`/`||` are already special operators and changing them to accept bool-ish doesn't mean we have to add implicit coercion to bool elsewhere (and indeed I wouldn't want to). However, I for one would find it unintuitive for the operators to accept bool-ish despite bool-ish not being a concept elsewhere in the language except in `if`. As far as I'm aware, other languages where these operators take bool-ish values also make use of bool-ish-ness in general. For example, [in Crystal "All values have an implicit truthiness" and the and/or operators return the last evaluated operand](https://crystal-lang.org/reference/1.18/tutorials/basics/50_control_flow.html#truthiness), not necessarily a boolean. [Similar in Python](https://docs.python.org/3/reference/expressions.html#boolean-operations) as was mentioned in the meeting.

**Comment by DanilaFe** [comment]
> As far as I'm aware, other languages where these operators take bool-ish values also make use of bool-ish-ness in general.

In what way do they make use of bool-ish-ness in general that is not related to `if` and `&&`? Right now I'd say we already have bool-ish-ness "in general" in the same way they do: it exists, but is only inspected by some language constructs.

**Comment by arifthpe** [comment]
> In what way do they make use of bool-ish-ness in general that is not related to if and &&? 

I guess "in general" is not very exact and I didn't have an exact idea in mind when I wrote that. Here's some ways I can think of:
- Defining the `!` operator on bool-ish inputs.
- Every value having a corresponding bool-ish value.
- `&&` being `boolish x boolish -> boolish` (rather than ` -> bool` in the Chapel proposal), which only makes sense as a consequence of every value having a bool-ish value.
- In Python specifically, being able to define the bool-ish value for class types via [`__bool__`](https://docs.python.org/3/reference/datamodel.html#object.__bool__), though I don't see that as very relevant here

**Comment by DanilaFe** [comment]
> && being boolish x boolish -> boolish (rather than  -> bool in the Chapel proposal), which only makes sense as a consequence of every value having a bool-ish value.

I don't think that `boolish x boolish -> boolish` is a signature in any of the languages, though I'm not sure if that affects your decision. Specifically,
* In languages with union types, the type is `A x B -> A | B`
* In languages without union types, the type is `boolish x boolish -> bool`

Since we don't have union types, we'd fit into the second category.

> In Python specifically, being able to define the bool-ish value for class types via [__bool__](https://docs.python.org/3/reference/datamodel.html#object.__bool__), though I don't see that as very relevant here

There is a path for this being possible for us (via the currently-internal `chpl_cond_test_method`). It's just out of scope for this proposal, I would say. FWIW I think there is precedent to use an interface here, but I digress.

> Every value having a corresponding bool-ish value.

This is the biggest distinction in Chapel, I'd say. That said, C/C++/ObjectiveC do not have a "general" truth value. E.g., tuples and other non-bool-coercible types are neither truthy nor falsey. Whether one should consider the C-derived rules a compelling precedent is a whole another issue 😉

**Comment by DanilaFe** [comment]
We held a non-binding vote on this during the public meeting. The results:

* ✅✅ ❌ (I think we should enable && for ints etc)
  * 5 thumbs up, one thinking emoji
* ✅ ❌❌ (I think we should keep the status quo)
  * one thumb up
* ❌❌❌ (I think we should disable int / classes for if, matching Swift and Rust)
  * 4 thumbs up, one heart, one thinking emoji

**Comment by bradcray** [comment]
Given the tight polling, I wanted to advocate here for the ✅✅ ❌ option on the grounds that it will require fewer changes to user code and our code combined with the fact that I don't believe the first ✅ has been a major pain point for Chapel users over its lifetime, and some would champion it (particularly those coming from C/C++ backgrounds, which is a lot of HPC users).

Put the opposite way, if we went with ❌❌❌, I could imagine a fair amount of grousing along the lines of "Is this change really worth my time and effort—aren't there more impactful things you could be solving or implementing?"

Though ✅ ❌❌ isn't my preference, I could see going with it just from the perspective of changing even less / avoiding a standoff between the two things that would involve changes.
